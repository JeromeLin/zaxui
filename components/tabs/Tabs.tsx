import React, { PureComponent } from 'react';
import TabBasePropsType from './PropsType';
import classnames from 'classnames';
import TabPanel from './TabPanel';
import TabHeader from './TabHeader';
import Carousel from '../carousel';


const getSelectIndex = (children) => {
  let selectIndex;
  React.Children.forEach(children, (item: any, index) => {
    if (item.props && item.props.selected) {
      selectIndex = index;
    }
  });
  return selectIndex;
};

export interface TabsProps extends TabBasePropsType {
  prefixCls?: string;
  className?: string;
}

export default class Tabs extends PureComponent<TabsProps, any> {
  static Panel: any;

  static defaultProps = {
    prefixCls: 'za-tabs',
    disabled: false,
    hasline: false,
    swipeable: false,
    defaultActiveKey:0,
    useTabPaged: false,
    horizontal: true,
    scrollElastic:true,
    activeKey:0
  };

  private carousel;


  constructor(props) {
    super(props);
    this.state = {
      activeKey: props.activeKey || props.defaultActiveKey || getSelectIndex(props.children) || 0,
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('activeKey' in nextProps || getSelectIndex(nextProps.children)) {
      this.setState({
        activeKey: nextProps.activeKey || nextProps.defaultActiveKey || getSelectIndex(nextProps.children) || 0,
      });
      if (typeof nextProps.onChange === 'function') {
        nextProps.onChange(nextProps.activeKey);
      }
    }
  }

  onSwipeChange = (activeKey) => {
    const { onChange  } = this.props;
    
    this.setState({ activeKey });
    if (typeof onChange === 'function') {
      onChange(activeKey);
    }
  }

  onTabClick = (tab, index) => {
    const { disabled, swipeable, onChange } = this.props;
    if (disabled || tab.props.disabled) {
      return;
    }
    this.setState({ activeKey: index });
    if (typeof onChange === 'function') {
      onChange(index);
    }
    if (swipeable) {
      this.carousel.onSlideTo(index);
    }
  }



  render() {
    const { prefixCls, className, hasline, swipeable, children, horizontal, useTabPaged ,page} = this.props;
    
    const classes = classnames(prefixCls, className,
      { [`${prefixCls}--hasline`]: hasline },
      { [`${prefixCls}--paged`]: page?false:useTabPaged },
      `${prefixCls}--${horizontal ? 'horizontal' : 'vertical'}`);
    // 渲染内容
    let contentRender;
    if (swipeable) {
      contentRender = (
        <Carousel
          direction={horizontal?"left":"right"}
          showPagination={false}
          activeIndex={this.state.activeKey}
          ref={(ele) => { this.carousel = ele; }}
          onChange={(activeKey) => this.onSwipeChange(activeKey)}
        >
          {React.Children.map(children, (item: any) => <div>{item.props.children}</div>)}
        </Carousel>
      );
    } else {
      contentRender = React.Children.map(children, (item: any, index) => {
        return <TabPanel {...item.props} selected={this.state.activeKey === index} />;
      });
    }
    return (
      <div className={classes}>
          <TabHeader
            {...this.props}
            activeIndex={this.state.activeKey}
            onTabHeaderClick={(tab, index) => { this.onTabClick(tab, index) }}
          ></TabHeader>
        <div className={`${prefixCls}__container`}>
          {contentRender}
        </div>
      </div>
    );
  }
}
