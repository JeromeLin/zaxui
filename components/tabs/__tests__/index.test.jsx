import React from 'react';
import { render, shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tabs from '../index';

const createPanel = (props, content, index) => {
  return (
    // eslint-disable-next-line no-undef
    <Tabs.Panel {...props} key={index}>
      <div >{content}</div>
    </Tabs.Panel>
  );
};
const createTabs = (props, childrenLen = []) => {
  return (
    <Tabs {...props}>
      {
        childrenLen.map((item, index) => {
          return createPanel({
            title: item.title,
          }, item.content, index);
        })
      }
    </Tabs>
  );
};

const createDefaultChild = (len) => {
  return Array.from({ length: len }, (v, k) => { return { title: `选项卡${k}`, content: `选项卡内容${k}` }; });
};

function createClientXYObject(x, y) {
  return { clientX: x, clientY: y };
}

function createStartTouchEventObject({ x = 0, y = 0, preventDefault = () => {} }) {
  return {
    touches: [createClientXYObject(x, y)],
    preventDefault,
  };
}

function createMoveTouchEventObject(props) {
  const { x = 0, y = 0, includeTouches = true, preventDefault = () => {} } = props;
  const moveTouchEvent = {
    changedTouches: [createClientXYObject(x, y)],
    preventDefault,
  };
  if (includeTouches) moveTouchEvent.touches = [createClientXYObject(x, y)];
  return moveTouchEvent;
}


describe('Tab', () => {
  // it('default selected', () => {
  //   const onChange = jest.fn();
  //   const wrapper = mount(
  //     <Tabs
  //       onChange={onChange}
  //       canSwipe={false}
  //       useTabPaged
  //     >
  //       <Tabs.Panel title="选项卡1">
  //         <div className="content">选项卡1内容</div>
  //       </Tabs.Panel>
  //       <Tabs.Panel title="选项卡2" isSelected={false}>
  //         <div className="content">选项卡2内容</div>
  //       </Tabs.Panel>
  //     </Tabs>
  //   );
  // });
  it('scrollElastic 禁止弹性 左侧右侧 ', () => {
    const onChange = jest.fn();
    const preventDefault = jest.fn();
    const props = { useTabPaged: true, canSwipe: true, scrollElastic: false };
    const wrapper = mount(
      createTabs({ ...props, onChange }, createDefaultChild(15))
    );
    // wrapper.find('.za-tabs__header').simulate('touchStart', {
    //   touches: [10, 0],
    // })
    //   .simulate('touchMove', {
    //     touches: [-400, 0],
    //   })
    //   .simulate('touchEnd', {
    //     touches: [200, 0],
    //   });
    wrapper.find('.za-tabs__header__item').first().simulate('touchStart', {
      touches: [10, 0],
    })
      .simulate('touchMove', createMoveTouchEventObject({ x: -400, y: 0, preventDefault }))
      .simulate('touchEnd', createMoveTouchEventObject({ x: 275, y: 0, preventDefault }));
    // wrapper.find('.za-tabs__header__item').first().simulate('touchEnd', {
    //   touches: [200, 0],
    // });
    wrapper.find('.za-tabs__header__item').first().simulate('click');
    wrapper.find('.za-tabs__header__item').last().simulate('click');
    expect(onChange).toBeCalledWith(0);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('tabWidth is Set', () => {
    const wrapper = shallow(
      createTabs({ tabWidth: 80 }, createDefaultChild(2))
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // it('useTabPaged', () => {
  //   const onChange = jest.fn();
  //   const props = { useTabPaged: true, canSwipe: true };
  //   const wrapper = mount(
  //     createTabs({ ...props, onChange }, createDefaultChild(15))
  //   );
  //   wrapper.find('.za-tabs__header__item').first().simulate('click');
  //   // expect(toJson(wrapper)).toMatchSnapshot();
  //   expect(onChange).toBeCalledWith(0);
  //   expect(toJson(wrapper)).toMatchSnapshot();
  // });
  // it('renders correctly', () => {
  //   const wrapper = render(
  //     createTabs({}, [{
  //       content: '试试点我左滑',
  //     },
  //     { content: '试试点我右滑' },
  //     ])
  //   );
  //   expect(toJson(wrapper)).toMatchSnapshot();
  // });

  // it('canSwipe', () => {
  //   const wrapper = render(
  //     createTabs({ canSwipe: true }, [{
  //       content: '试试点我左滑',
  //     },
  //     { content: '试试点我右滑' },
  //     ])
  //   );
  //   expect(toJson(wrapper)).toMatchSnapshot();
  // });

  // it('canNotSwipe', () => {
  //   const wrapper = render(
  //     createTabs({ canSwipe: false }, [{
  //       content: '试试点我左滑',
  //     },
  //     { content: '试试点我右滑' },
  //     ])
  //   );
  //   expect(toJson(wrapper)).toMatchSnapshot();
  // });

  it('lineWidth is auto', () => {
    const wrapper = shallow(
      createTabs({ lineWidth: 30 }, [{
        content: '试试点我左滑',
      },
      { content: '试试点我右滑' },
      ])
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    // wrapper.setProps({ lineWidth: 30 });
    // expect(toJson(wrapper)).toMatchSnapshot();
  });

  // it('receive new value', () => {
  //   const onChange = jest.fn();
  //   const wrapper = shallow(
  //     createTabs({ onChange }, [{
  //       content: '试试点我左滑',
  //     },
  //     { content: '试试点我右滑' },
  //     ])
  //   );
  //   wrapper.setProps({ value: 1 });
  // });

  // it('click tabs', () => {
  //   const onChange = jest.fn();
  //   const wrapper = mount(
  //     <Tabs canSwipe onChange={onChange}>
  //       <Tabs.Panel title="选项卡1" disabled>
  //         <div>试试点我左滑</div>
  //       </Tabs.Panel>
  //       <Tabs.Panel title="选项卡2">
  //         <div>试试点我右滑</div>
  //       </Tabs.Panel>
  //     </Tabs>
  //   );
  //   wrapper.find('.za-tabs__header__item').first().simulate('click');
  //   wrapper.find('.za-tabs__header__item').last().simulate('click');
  //   expect(onChange).toBeCalledWith(1);
  // });
});
