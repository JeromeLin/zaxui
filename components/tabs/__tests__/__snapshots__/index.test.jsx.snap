// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab lineWidth is auto 1`] = `
<div
  className="za-tabs za-tabs--horizontal"
>
  <isPrevTabHeader
    activeIndex={0}
    animationDuration={300}
    canSwipe={false}
    defaultValue={0}
    disabled={false}
    hasline={false}
    horizontal={true}
    lineWidth={30}
    moveDistanceRatio={0.5}
    onTabHeaderClick={[Function]}
    page={5}
    prefixCls="za-tabs"
    scrollElastic={true}
    tabHeight={40}
    tabWidth={70}
    useTabPaged={false}
  >
    <TabPanel
      key="0"
      prefixCls="za-tabs"
    >
      <div>
        试试点我左滑
      </div>
    </TabPanel>
    <TabPanel
      key="1"
      prefixCls="za-tabs"
    >
      <div>
        试试点我右滑
      </div>
    </TabPanel>
  </isPrevTabHeader>
  <div
    className="za-tabs__container"
  >
    <TabPanel
      key=".$0"
      prefixCls="za-tabs"
      selected={true}
    >
      <div>
        试试点我左滑
      </div>
    </TabPanel>
    <TabPanel
      key=".$1"
      prefixCls="za-tabs"
      selected={false}
    >
      <div>
        试试点我右滑
      </div>
    </TabPanel>
    0
  </div>
</div>
`;

exports[`Tab scrollElastic 禁止弹性 左侧右侧  1`] = `
<Tabs
  canSwipe={true}
  defaultValue={0}
  disabled={false}
  hasline={false}
  horizontal={true}
  moveDistanceRatio={0.5}
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          0,
        ],
        Array [
          14,
        ],
        Array [
          14,
        ],
        Array [
          14,
        ],
        Array [
          14,
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  page={5}
  prefixCls="za-tabs"
  scrollElastic={false}
  useTabPaged={true}
>
  <div
    className="za-tabs za-tabs--paged za-tabs--horizontal"
  >
    <isPrevTabHeader
      activeIndex={14}
      animationDuration={300}
      canSwipe={true}
      defaultValue={0}
      disabled={false}
      hasline={false}
      horizontal={true}
      moveDistanceRatio={0.5}
      onChange={
        [MockFunction] {
          "calls": Array [
            Array [
              0,
            ],
            Array [
              14,
            ],
            Array [
              14,
            ],
            Array [
              14,
            ],
            Array [
              14,
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
      onTabHeaderClick={[Function]}
      page={5}
      prefixCls="za-tabs"
      scrollElastic={false}
      tabHeight={40}
      tabWidth={70}
      useTabPaged={true}
    >
      <div
        className="za-tabs__header"
      >
        <Drag
          onDragEnd={[Function]}
          onDragMove={[Function]}
        >
          <ul
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tablist"
          >
            <li
              className="za-tabs__header__item"
              key=".$0"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡0
            </li>
            <li
              className="za-tabs__header__item"
              key=".$1"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡1
            </li>
            <li
              className="za-tabs__header__item"
              key=".$2"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡2
            </li>
            <li
              className="za-tabs__header__item"
              key=".$3"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡3
            </li>
            <li
              className="za-tabs__header__item"
              key=".$4"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡4
            </li>
            <li
              className="za-tabs__header__item"
              key=".$5"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡5
            </li>
            <li
              className="za-tabs__header__item"
              key=".$6"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡6
            </li>
            <li
              className="za-tabs__header__item"
              key=".$7"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡7
            </li>
            <li
              className="za-tabs__header__item"
              key=".$8"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡8
            </li>
            <li
              className="za-tabs__header__item"
              key=".$9"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡9
            </li>
            <li
              className="za-tabs__header__item"
              key=".$10"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡10
            </li>
            <li
              className="za-tabs__header__item"
              key=".$11"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡11
            </li>
            <li
              className="za-tabs__header__item"
              key=".$12"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡12
            </li>
            <li
              className="za-tabs__header__item"
              key=".$13"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡13
            </li>
            <li
              className="za-tabs__header__item"
              key=".$14"
              onClick={[Function]}
              role="tab"
              style={
                Object {
                  "flexBasis": 70,
                  "height": "auto",
                }
              }
            >
              选项卡14
            </li>
          </ul>
        </Drag>
        <div
          className="za-tabs__line"
          style={
            Object {
              "width": 70,
            }
          }
        />
      </div>
    </isPrevTabHeader>
    <div
      className="za-tabs__container"
    >
      <Carousel
        activeIndex={14}
        animationDuration={300}
        autoPlay={false}
        autoPlayIntervalTime={3000}
        direction="left"
        height={160}
        loop={false}
        moveDistanceRatio={0.5}
        moveTimeSpan={300}
        onChange={[Function]}
        prefixCls="za-carousel"
        showPagination={false}
      >
        <div
          className="za-carousel za-carousel--horizontal"
        >
          <Drag
            onDragEnd={[Function]}
            onDragMove={[Function]}
            onDragStart={[Function]}
          >
            <div
              className="za-carousel__items"
              onMouseDown={[Function]}
              onMouseMove={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              onTransitionEnd={[Function]}
              style={Object {}}
            >
              <div
                className="za-carousel__item"
                key="0/.$.$0"
              >
                <div>
                  选项卡内容0
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="1/.$.$1"
              >
                <div>
                  选项卡内容1
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="2/.$.$2"
              >
                <div>
                  选项卡内容2
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="3/.$.$3"
              >
                <div>
                  选项卡内容3
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="4/.$.$4"
              >
                <div>
                  选项卡内容4
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="5/.$.$5"
              >
                <div>
                  选项卡内容5
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="6/.$.$6"
              >
                <div>
                  选项卡内容6
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="7/.$.$7"
              >
                <div>
                  选项卡内容7
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="8/.$.$8"
              >
                <div>
                  选项卡内容8
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="9/.$.$9"
              >
                <div>
                  选项卡内容9
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="10/.$.$10"
              >
                <div>
                  选项卡内容10
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="11/.$.$11"
              >
                <div>
                  选项卡内容11
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="12/.$.$12"
              >
                <div>
                  选项卡内容12
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="13/.$.$13"
              >
                <div>
                  选项卡内容13
                </div>
              </div>
              <div
                className="za-carousel__item"
                key="14/.$.$14"
              >
                <div>
                  选项卡内容14
                </div>
              </div>
            </div>
          </Drag>
        </div>
      </Carousel>
      14
    </div>
  </div>
</Tabs>
`;

exports[`Tab tabWidth is Set 1`] = `
<div
  className="za-tabs za-tabs--horizontal"
>
  <isPrevTabHeader
    activeIndex={0}
    animationDuration={300}
    canSwipe={false}
    defaultValue={0}
    disabled={false}
    hasline={false}
    horizontal={true}
    moveDistanceRatio={0.5}
    onTabHeaderClick={[Function]}
    page={5}
    prefixCls="za-tabs"
    scrollElastic={true}
    tabHeight={40}
    tabWidth={80}
    useTabPaged={false}
  >
    <TabPanel
      key="0"
      prefixCls="za-tabs"
      title="选项卡0"
    >
      <div>
        选项卡内容0
      </div>
    </TabPanel>
    <TabPanel
      key="1"
      prefixCls="za-tabs"
      title="选项卡1"
    >
      <div>
        选项卡内容1
      </div>
    </TabPanel>
  </isPrevTabHeader>
  <div
    className="za-tabs__container"
  >
    <TabPanel
      key=".$0"
      prefixCls="za-tabs"
      selected={true}
      title="选项卡0"
    >
      <div>
        选项卡内容0
      </div>
    </TabPanel>
    <TabPanel
      key=".$1"
      prefixCls="za-tabs"
      selected={false}
      title="选项卡1"
    >
      <div>
        选项卡内容1
      </div>
    </TabPanel>
    0
  </div>
</div>
`;
